File: packages/gbnf-experiment/data/20260909T002206Z_273cf6b9/ported_implementation/tests/unit/grammar-graph/get-serialized-rule-key.test.ts
import { beforeEach, describe, expect, test, vi } from 'vitest';
import {
  KEY_TRANSLATION,
  getSerializedRuleKey,
} from '../../../src/grammar-graph/get-serialized-rule-key.ts';
import {
  RuleChar,
  RuleCharExclude,
  RuleEnd,
} from '../../../src/grammar-graph/grammar-graph-types.ts';
import { RuleRef } from '../../../src/grammar-graph/rule-ref.ts';
import {
  isRuleChar,
  isRuleCharExclude,
  isRuleEnd,
  isRuleRef,
} from '../../../src/grammar-graph/type-guards.ts';

vi.mock('../../../src/grammar-graph/type-guards.ts', async importOriginal => {
  const actual = await importOriginal<
    typeof import('../../../src/grammar-graph/type-guards.ts')
  >();
  return {
    ...actual,
    isRuleEnd: vi.fn(() => false),
    isRuleChar: vi.fn(() => false),
    isRuleCharExclude: vi.fn(() => false),
    isRuleRef: vi.fn(() => false),
  };
});

beforeEach(() => {
  for (const guard of [isRuleEnd, isRuleChar, isRuleCharExclude, isRuleRef]) {
    vi.mocked(guard).mockReturnValue(false);
  }
});

describe('get_serialized_rule_key', () => {
  test('returns type for end rules', () => {
    vi.mocked(isRuleEnd).mockReturnValue(true);
    expect(getSerializedRuleKey(new RuleEnd())).toBe(`${KEY_TRANSLATION.RuleEnd}`);
  });

  test('returns type and value for character rules', () => {
    vi.mocked(isRuleChar).mockReturnValue(true);
    expect(getSerializedRuleKey(new RuleChar([97]))).toBe(`${KEY_TRANSLATION.RuleChar}-[97]`);
  });

  test('returns type and value for character exclude rules', () => {
    vi.mocked(isRuleCharExclude).mockReturnValue(true);
    expect(getSerializedRuleKey(new RuleCharExclude([97]))).toBe(
      `${KEY_TRANSLATION.RuleCharExclude}-[97]`,
    );
  });

  test('returns ref type with value for reference rules', () => {
    vi.mocked(isRuleRef).mockReturnValue(true);
    expect(getSerializedRuleKey(new RuleRef(99))).toBe('3-99');
  });

  test('throws error for unknown rule types', () => {
    const rule = { type: 'UNKNOWN', value: 'something' };
    expect(() => getSerializedRuleKey(rule as never)).toThrow(/Unknown rule type/);
  });
});

File: proposed-projects/round-trip-experiment/reverse/20260909T002206Z_273cf6b9/20260910T042804Z_e72f6c04/ported_implementation/gbnf/grammar_graph/get_serialized_rule_key_test.py
import pytest

from . import get_serialized_rule_key as module
from .get_serialized_rule_key import KEY_TRANSLATION, get_serialized_rule_key
from .grammar_graph_types import RuleChar, RuleCharExclude, RuleEnd
from .rule_ref import RuleRef

GUARDS = ["is_rule_end", "is_rule_char", "is_rule_char_exclude", "is_rule_ref"]


@pytest.fixture(autouse=True)
def stub_guards(monkeypatch):
    """Every guard reports False unless a test opts one back in."""
    for guard in GUARDS:
        monkeypatch.setattr(module, guard, lambda rule: False)
    yield monkeypatch


def enable(monkeypatch, guard):
    monkeypatch.setattr(module, guard, lambda rule: True)


def test_returns_type_for_end_rules(stub_guards):
    enable(stub_guards, "is_rule_end")
    assert get_serialized_rule_key(RuleEnd()) == f"{KEY_TRANSLATION.RuleEnd}"


def test_returns_type_and_value_for_character_rules(stub_guards):
    enable(stub_guards, "is_rule_char")
    assert (
        get_serialized_rule_key(RuleChar([97])) == f"{KEY_TRANSLATION.RuleChar}-[97]"
    )


def test_returns_type_and_value_for_character_exclude_rules(stub_guards):
    enable(stub_guards, "is_rule_char_exclude")
    assert (
        get_serialized_rule_key(RuleCharExclude([97]))
        == f"{KEY_TRANSLATION.RuleCharExclude}-[97]"
    )


def test_returns_ref_type_with_value_for_reference_rules(stub_guards):
    enable(stub_guards, "is_rule_ref")
    assert get_serialized_rule_key(RuleRef(99)) == "3-99"


def test_raises_for_unknown_rule_types():
    with pytest.raises(ValueError, match="Unknown rule type"):
        get_serialized_rule_key({"type": "UNKNOWN", "value": "something"})
